### DISCUSSION TITLE
Start Langium as standalone server

### DISCUSSION THREAD

**Discussion by theogiraudet** [eclipse-langium/langium#471] [Q&A] [answered]
Hi!
I want to use my DSL in others IDEs using Monaco, how can I run the Langium server as a standalone program?
I tried to launch the main in these two ways:
```
node .\out\language-server\main.js --node-ipc
node .\out\language-server\main.js --stdio
```
But it doesn't seem to work.

Thanks!

**Comment by msujew** [comment]
Hi @theogiraudet,

That's weird, I just tried to use `node .\out\language-server\main.js --stdio`, which seems to work fine for me. What issues are you experiencing with that approach? Any errors or something similar in the console?

**Reply by theogiraudet** [reply]
Um... Maybe I do not have the right approach... No, no error, just the proccess running (I guess), but no message in the console to confirm that. I think is more "--socket" I need, I want to connect my server to a Monaco (I already have a working front).
I tried `node .\out\language-server\main.js --socket=20000` but the program stop immediatly (no errors, no logs). No process runs on the port 20000 either.

**Reply by msujew** [reply]
Yeah, that sounds correct (for `--stdio`). The issue is that the language server isn't allowed to print anything to the console, since the console is the transport medium. Anything that is printed to the console gets interpreted by the language client. It only prints valid LSP responses (after receiving valid LSP requests).

> I tried `node .\out\language-server\main.js --socket=20000` but the program stop immediatly (no errors, no logs).

This one is a bit tricky, and I had to look into the [language server source code](https://github.com/microsoft/vscode-languageserver-node/blob/48c9f063483c78b99b988b78f517b24106be886c/server/src/node/main.ts#L207-L209) to find a solution: The `--socket` and actual port argument need to be their own arguments, not a single one. This should work:

```
node .\out\language-server\main.js --socket 20000
```

**Reply by theogiraudet** [reply]
Thanks for the details about stdio!
I think the two solutions work (usually), if I check the source code:
![image](https://user-images.githubusercontent.com/28688370/162728521-505c6817-11a7-4955-a6e0-e4ed103c14da.png)
But in these two cases, no process runs on the port 2000.

**Reply by msujew** [reply]
Hm, true, that's weird. Doesn't work for me either...

I digged a bit more into it, and it seems like the `--socket` option works a bit ... let's say unexpected. See [this explanation by the devs](https://github.com/microsoft/vscode-languageserver-node/issues/245#issuecomment-336136082). Instead of using the language server as the actual server (i.e. listening to the specified port), the language server expects the client to act as the server, to which it will try to connect to.

I assume that since no client provides the specified port on your machine, the language server simply ends the process, as there's no one to connect to.

**Reply by theogiraudet** [reply]
Thanks for the link and for the response!
For the moment, even with the code I can't do it, I'm still looking.
Perhaps it would be a good idea to propose with Langium a way to create sockets more easily or do you think it's out of the goal of the project?

**Reply by spoenemann** [reply]
I think that is out of scope for Langium. You can use Node.js to create a socket server and connect the language server to it using the `vscode-languageserver` API. It would be nice to have an example for that, but so far I'm not aware of any.

**Reply by theogiraudet** [reply]
Hi,
Thanks you for your answer. I built a bridge opening a Websocket and connecting to Langium via stdout and it's work perfectly.
I don't really know if it's a perfect solution but it's enough for my use case !

**Comment by spoenemann** [comment, marked as answer]
You can connect Langium directly to a WebSocket with a custom message reader / writer. The following snippet is based on the `ws` package:

```typescript
class WebSocketMessageWriter extends AbstractMessageWriter {
    private readonly ws: WebSocket;

    constructor(ws: WebSocket) {
        super();
        this.ws = ws;
    }

    write(msg: Message): Promise<void> {
        return new Promise((resolve, reject) => {
            const content = JSON.stringify(msg);
            this.ws.send(content, err => {
                if (err) {
                    reject(err);
                } else {
                    resolve();
                }
            });
        });
    }

    end(): void {}
}

class WebSocketMessageReader extends AbstractMessageReader {
    private readonly ws: WebSocket;

    constructor(ws: WebSocket) {
        super();
        this.ws = ws;
    }

    listen(callback: DataCallback): Disposable {
        const messageListener = (rawMessage: Message) => {
            const message = JSON.parse(rawMessage.toString()) as Message;
            callback(message);
        };
        const closeListener = () => this.fireClose()

        this.ws.on('message', messageListener);
        this.ws.on('close', closeListener);

        return {
            dispose: () => {
                this.ws.removeListener('message', messageListener);
                this.ws.removeListener('close', closeListener);
            }
        };
    }
}

const wsServer = new WebSocket.Server(/* options */);
wsServer.on('connection', (ws, request) => {
    const reader = new WebSocketMessageReader(ws);
    const writer = new WebSocketMessageWriter(ws);
    const connection = createConnection(reader, writer);

    const services = createMyLanguageServices({ connection });
    startLanguageServer(services.shared);
});
```

**Reply by theogiraudet** [reply]
I have replaced my bridge with your implementations and it works perfectly, thanks!
